### PR TITLE
Filling in functionality + tests

### PR DESCRIPTION
**Pull request by kellysutton**
This is not as DRY as it should be, ideally. There are also a few remaining failing tests.

### CODE REVIEW

**Comment by kellysutton** [comment]
This is coming along, but I'm getting tripped up on using fully-qualified URLs as path components. The `net/url` package QueryEscape function is insufficient. We need something akin to the Ruby `CGI::escape` method. Source of that method pasted below.

```ruby
def CGI::escape(string)
  string.gsub(/([^ a-zA-Z0-9_.-]+)/) do
    '%' + $1.unpack('H2' * $1.bytesize).join('%').upcase
  end.tr(' ', '+')
end
```

Any ideas? cc @parkr

File: imgix.go [hunk 1]
```diff
@@ -74,22 +80,46 @@ func (c *Client) Host(path string) string {
...
+		Path:     imgPath,
+		RawQuery: params.Encode(),
+	}
+
+	signature := c.SignatureForPathAndParams(imgPath, params)
+	if signature != "" && len(params) > 0 {
+		u.RawQuery = u.RawQuery + "&" + signature
+	} else if signature != "" && len(params) == 0 {
+		u.RawQuery = signature
+	}
```
Comment (parkr) [inline imgix.go]: what do you think about:

```go
params.Add("s", c.SignatureForPathAndParams(imgPath, params))

u := url.Url{
    // ...
    RawQuery: params.Encode()
}
return u.String()
```

File: imgix_test.go [hunk 2]
```diff
@@ -32,8 +36,37 @@ func TestClientScheme(t *testing.T) {
 
 func TestClientURL(t *testing.T) {
```
Comment (parkr) [inline imgix_test.go]: Can you rename this test, too?

File: imgix.go (hunk 1 above)
Comment (kellysutton) [inline imgix.go]: Signature will be `""` when a token is not given to the client. It's possible to use unsigned URLs, although we don't recommend it.

It's also important that the signature always be the **last** parameter when present, hence the gross string concatenation. I'll leave a comment about that.

File: imgix_test.go (hunk 2 above)
Comment (kellysutton) [inline imgix_test.go]: Yep, will do.

**Comment by parkr** [comment]
Applying this diff:

```diff
diff --git a/imgix.go b/imgix.go
index 6645a68..f4d1184 100644
--- a/imgix.go
+++ b/imgix.go
@@ -16,7 +16,7 @@ const (
 	ShardStrategyCycle = ShardStrategy(":cycle")
 )
 
-var RegexpRemoveHTTPAndS = regexp.MustCompile("http(s)://")
+var RegexpHTTPAndS = regexp.MustCompile("http(s)?://")
 
 func NewClient(hosts ...string) Client {
 	return Client{hosts: hosts, secure: true}
@@ -77,7 +77,7 @@ func (c *Client) Host(path string) string {
 		c.cycleIndex = (c.cycleIndex + 1) % len(c.hosts)
 	}
 
-	return RegexpRemoveHTTPAndS.ReplaceAllString(host, "")
+	return RegexpHTTPAndS.ReplaceAllString(host, "")
 }
 
 func (c *Client) SignatureForPath(path string) string {
@@ -106,14 +106,21 @@ func (c *Client) Path(imgPath string) string {
 }
 
 func (c *Client) PathWithParams(imgPath string, params url.Values) string {
+	var path string
+	if RegexpHTTPAndS.MatchString(imgPath) {
+		path = url.QueryEscape(imgPath)
+	} else {
+		path = imgPath
+	}
+
 	u := url.URL{
 		Scheme:   c.Scheme(),
 		Host:     c.Host(imgPath),
-		Path:     imgPath,
+		Path:     path,
 		RawQuery: params.Encode(),
 	}
 
-	signature := c.SignatureForPathAndParams(imgPath, params)
+	signature := c.SignatureForPathAndParams(path, params)
 	if signature != "" && len(params) > 0 {
 		u.RawQuery = u.RawQuery + "&" + signature
 	} else if signature != "" && len(params) == 0 {
```

encoded the path properly... it was just different from the expected output because it also stuck a bunch of `25`'s in there, like `%253A` instead of `%3A`.

**Comment by kellysutton** [comment]
Yeah, the `url.QueryEscape` function is deficient for this reason, as it looks like it double-encodes everything.

A colon should just become `"%3A"`. The `url.QueryEscape` function is taking a second pass on that and escaping the percent sign that it generated, making a colon `"%253A"`. I'll ping some folks internally to see if there's a better way to handle this.

**Comment by nf** [comment]
`url.QueryEscape` escapes unescaped strings. If you pass in an escaped query string then it'll double escape it. There's no way not to do this. Correct me if I'm misunderstanding the issue here.

**Comment by nf** [comment]
You should *not* be query escaping a string that you put into a `URL.Path` field.

**Comment by kellysutton** [comment]
Correct, we need to be CGI-escaping the URL given. For imgix to be able to reliably proxy fully-qualified URLs and maintain any URL parameters in the origin URL, we have to escape the entire origin URL. See this section of our blueprint doc for more info: https://github.com/imgix/imgix-blueprint#fully-qualified-urls

Does that make sense?

**Comment by nf** [comment]
I think you should put the unescaped path in the URL.Path field.
If that doesn't do what you want, then you should construct the URL by hand.

**Comment by kellysutton** [comment]
Noted. Thanks for the help :+1:

**Comment by parkr** [comment]
Thanks, @nf!

**Comment by parkr** [comment]
> If that doesn't do what you want, then you should construct the URL by hand.

@kellysutton Is this what you want to attempt?

**Comment by kellysutton** [comment]
@parkr Yeah, that's what we'll have to do in this case. I'll adapt the Ruby code and give it a whirl.

**Pull request merged**